### DISCUSSION TITLE
flash attention in tinygrad

### DISCUSSION THREAD

**Discussion by vish-pr** [tinygrad/tinygrad#2955] [General]
Is there any implementation existing? Or someone writing it?

**Comment by girldeath** [comment]
Flash attention is two things:
1. Attention with sub-quadratic memory accesses, because of online softmax
2. An optimized/fused kernel for attention

You can implement the first, without Tinygrad needing to do anything.
The second needs Tinygrad to handle kernel fusion. As far as I understand, flash attention will just happen when all the requisite kernel fusions are supported. At that point, a whole class of optimizations (which flash attention is part of) would be possible.
